build.gradle

```diff
@@ -324,3 +328,78 @@ tasks.withType(Test) {
         }
     }
 }
+
+
+task checkForMissingRunConfigurations() {
+    def runConfigurationsDir = new File(".idea/runConfigurations")
+    if (!runConfigurationsDir.exists()) {
+        generateRunConfigurations()
+    }
+}
+
+task regenerateRunConfigurations() {
+    doLast {
+        generateRunConfigurations()
+    }
+}
+
+private void generateRunConfigurations() {
+
+
+
+    createGradleRunConfiguration('Run IDE (default)', 'runIde', '', '')
+    createGradleRunConfiguration('Run Tests (default)', 'clean test', '', '')
+
+    createGradleRunConfiguration('Generate Parsers', 'cleanGenerated generateHaxeParser generateHxmlParser', '', '')
+
+
+    createGradleRunConfiguration('Run Tests 2018', 'clean test', "-PtargetVersion=$latest2018Version", '')
+    createGradleRunConfiguration('Run Tests 2017', 'clean test', "-PtargetVersion=$latest2017Version", '')
+    createGradleRunConfiguration('Run Tests 2016', 'clean test', "-PtargetVersion=$latest2016Version", '')
```

Comment:
I'm not sure about needing versioned run targets.  This will run the currently built project (built with whatever IDEA you are running) against a specific version of IDEA, right?  So, running the 2017.2 target from 2018.1 won't work because of the incompatibilities, correct?  Or am I getting that wrong and the 2017.2 version of IDEA will be used to build prior to the tests being run, so that the environments will be consistent?

Reply:
there should not be any incompatibilities as far as i know (2016, 2017, 2018 all works in my 2017.3 install, but if you find its cluttering the project i could remove them).

The build command will use the  javac2 compiler, sources and libs from the version of intelliJ that the configuration specify (it will download it if its missing), so it should be pretty close if not identical to what you would get if you compiled it with an installed version of intelliJ with the same version number.

if you build using the `--debug`  flag you will see that version specific files that are resolved for the build.

2016  runConfiguration
```
Build operation 'Resolve lib/idea.jar (com.jetbrains:ideaIU:2016.3.8)' completed
Build operation 'Resolve lib/idea_rt.jar (com.jetbrains:ideaIU:2016.3.8)' completed
Build operation 'Resolve lib/ideax.jar (com.jetbrains:ideaIU:2016.3.8)' completed
Build operation 'Resolve lib/eddsa-0.2.0.jar (com.jetbrains:ideaIU:2016.3.8)' started
....
Build operation 'Resolve lib/javac2.jar (com.jetbrains:ideaIU:2016.3.8)' started 
```

2017  runConfiguration
```
Resolve lib/jasper21_rt.jar (com.jetbrains:ideaIU:2017.3)' started
Resolve lib/jdkAnnotations.jar (com.jetbrains:ideaIU:2017.3)' started
Resolve lib/jasper2_rt.jar (com.jetbrains:ideaIU:2017.3)' started
Resolve lib/idea.jar (com.jetbrains:ideaIU:2017.3)' completed
....
'Resolve lib/javac2.jar (com.jetbrains:ideaIU:2017.3)' started`
```

The plugin also creates a sandbox for the version you provide and starts and instance of that version with the plugin installed, so the only incompatibility i can think of is older versions not supporting java 8 or something like that.

Reply:
OK, then.  Great.